### PR TITLE
TRCFileAdapter now reads blanks as NaN and does not use space as delimiter

### PR DESCRIPTION
**Pull request by aseth1**
Fixes issue #2148, #1974 

Ran into this is in trying to test our C3D reader by comparing markers outputted as a TRC file with Mokka export of C3D data as TRC. Mokka outputs blank delimited by tabs to denote obstructed or out of view marker.

### Brief summary of changes
- revised `FileAdapter::tokenize()` to read consecutive tabs as empty strings
- removed space as a delimiter
- empty strings automatically convert to NaN
- add test case using Mokka generated TRC
- renamed data files for testing based on the conditions they are testing

### Testing I've completed
- tests pass locally
- added typical file with blank entries (from Mokka) 

### CHANGELOG.md (choose one)
- no need to update because this is a bug fix

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

### CODE REVIEW

File: OpenSim/Common/Test/testTRCFileAdapter.cpp [hunk 1]
```diff
@@ -60,77 +40,130 @@ void testFailed(const std::string& filename,
...
-                testFailed(filenameA, tokenA, tokenB);
-            }
 
-            if(d_tokenA != d_tokenB && 
-               !(std::isnan(d_tokenA) && 
-                 std::isnan(d_tokenB))) {
-                testFailed(filenameA, tokenA, tokenB);
+    // Use a STOFileAdapter just to access the common
+    // FileAdapter utilities.
+    OpenSim::STOFileAdapter sfa;
```
Comment (chrisdembia) [inline OpenSim/Common/Test/testTRCFileAdapter.cpp]: By using `FileAdapter` utilities to to test the file adapters, is it possible to have a bug in `getNextLine()` and the tests would still pass?

File: OpenSim/Common/Test/testTRCFileAdapter.cpp [hunk 2]
```diff
@@ -60,77 +40,130 @@ void testFailed(const std::string& filename,
...
+            if (tokensA[i] != tokensB[i]) {
+                std::string tokenA{ tokensA[i] };
+                std::string tokenB{ tokensB[i] };
+                OpenSim::IO::TrimWhitespace(tokenA);
+                OpenSim::IO::TrimWhitespace(tokenB);
+                // We interpreted blank as NaN now revert
+                // to compare to the original file with blanks
+                if (tokenB == "nan") {
+                    tokenB = "";
+                }
```
Comment (chrisdembia) [inline OpenSim/Common/Test/testTRCFileAdapter.cpp]: Why do you check only `tokenB`?

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 3]
```diff
@@ -162,12 +192,16 @@ TRCFileAdapter::extendRead(const std::string& fileName) const {
...
-            row_vector{static_cast<int>(num_markers_expected)};
+            row_vector{static_cast<int>(num_markers_expected), 
+                       SimTK::Vec3(SimTK::NaN)};
         int ind{0};
-        for(std::size_t c = 2; c < column_labels.size() * 3 + 2; c += 3)
-            row_vector[ind++] = SimTK::Vec3{std::stod(row.at(c)),
-                                            std::stod(row.at(c+1)),
-                                            std::stod(row.at(c+2))};
+        for (std::size_t c = 2; c < column_labels.size() * 3 + 2; c += 3) {
+            if (!row.at(c).empty())
```
Comment (chrisdembia) [inline OpenSim/Common/TRCFileAdapter.cpp]: Can you use `{}` for this `if`? That'll help make it clearer that you want `++ind` outside the `if`.

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 4]
```diff
@@ -162,12 +192,16 @@ TRCFileAdapter::extendRead(const std::string& fileName) const {
 
         // Columns 2 till the end are data.
         TimeSeriesTableVec3::RowVector 
-            row_vector{static_cast<int>(num_markers_expected)};
+            row_vector{static_cast<int>(num_markers_expected), 
+                       SimTK::Vec3(SimTK::NaN)};
         int ind{0};
-        for(std::size_t c = 2; c < column_labels.size() * 3 + 2; c += 3)
-            row_vector[ind++] = SimTK::Vec3{std::stod(row.at(c)),
```
Comment (chrisdembia) [inline OpenSim/Common/TRCFileAdapter.cpp]: What if `row.at(c+1)` or `row.at(c+2)` is empty?

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 5]
```diff
@@ -150,7 +174,13 @@ TRCFileAdapter::extendRead(const std::string& fileName) const {
     // the data container.
     std::size_t line_num{_dataStartsAtLine - 1};
     auto row = nextLine();
+
     while(!row.empty()) {
+        if (row.at(0).empty()) { //if no Frame# then ignore as empty elements
```
Comment (chrisdembia) [inline OpenSim/Common/TRCFileAdapter.cpp]: What does "ignore as empty elements" mean?

What's the reason for this change? Isn't possible for a frame number to be missing but the rest of the row has valid data?

It would be good to document details like this in the class description for `TRCFileAdapter`, so a user can troubleshoot parsing issues.

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 6]
```diff
@@ -103,6 +120,10 @@ TRCFileAdapter::extendRead(const std::string& fileName) const {
     // Read the line containing column labels and fill up the column labels
     // container.
     auto column_labels = nextLine();
+    // for marker labels we do not need three columns per marker.
+    // remove the blank ones used in TRC due to tabbing
```
Comment (chrisdembia) [inline OpenSim/Common/TRCFileAdapter.cpp]: It would be great if the comment could show a diagram or TRC snippet. The point here is that the TRC file looks like the following:

```
marker1<tab><tab><tab>marker2<tab><tab><tab>...
X1<tab>Y1<tab>Y2<tab>X2<tab>Y2<tab>Z2<tab>...
```

You're removing empty elements introduced by the tabs in the first of those rows, correct?

File: OpenSim/Common/FileAdapter.h [hunk 7]
```diff
@@ -201,18 +201,18 @@ class OSIMCOMMON_API FileAdapter : public DataAdapter {
     static
     std::string findExtension(const std::string& filename);
 
+    /** Get the next line from the stream and tokenize/split the line using
+    the given delimiters.                                                     */
+    std::vector<std::string> getNextLine(std::istream& stream,
+        const std::string& delims) const;
```
Comment (chrisdembia) [inline OpenSim/Common/FileAdapter.h]: I think this could be static.

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 8]
```diff
@@ -133,6 +154,9 @@ TRCFileAdapter::extendRead(const std::string& fileName) const {
     // X1, Y1, Z1, X2, Y2, Z2, ... so on.
     // Check and ignore these labels.
     auto xyz_labels_found = nextLine();
+    // erase blank labels, e.g. due to Frame# and Time columns
+    eraseEmptyElements(xyz_labels_found);
```
Comment (chrisdembia) [inline OpenSim/Common/TRCFileAdapter.cpp]: Just so I understand: the reason the calls to `eraseEmptyElemnts()` is necessary now is because space is no longer a delimiter?

File: OpenSim/Common/FileAdapter.cpp [hunk 9]
```diff
@@ -53,27 +53,19 @@ FileAdapter::tokenize(const std::string& str,
...
-                is_token = false;
-            }
-        } else {
-            if(!is_token) {
-                token_start = token_end;
-                is_token = true;
-            }
+    size_type token_start{0}, token_end{ std::string::npos };
+    while((token_end = str.find_first_of(delims, token_start)) != std::string::npos) {
+        if (token_end >= token_start) {
```
Comment (chrisdembia) [inline OpenSim/Common/FileAdapter.cpp]: Is it possible for `token_end < token_start`?

**Comment by chrisdembia** [comment]
Oops; I didn't notice this PR was WIP. Sorry if you didn't want comments yet.

**Comment by aseth1** [comment]
@chrisdembia no worries about the early comments. They were helpful in my ongoing revisions.  The PR is ready for review. I did end up having some modest impact the `DelimFileAdapter` and `STOFileAdapter` and believe behavior is more consistent. I did make `FileAdapter::getNextLine()` static and some other cleanup to reduce code duplication. I also added another data file that combines blanks and NaNs. Coverage is not yet complete, but I believe it is stricter and covers more than before.

File: OpenSim/Common/FileAdapter.cpp (hunk 9 above)
Comment (aseth1) [reply OpenSim/Common/FileAdapter.cpp]: Yes, because after a token is processed, `token_start = token_end` and immediately thereafter we increment `++token_start`, in which case it is now greater than `token_end`.

File: OpenSim/Common/DelimFileAdapter.h [hunk 10]
```diff
@@ -370,10 +371,17 @@ DelimFileAdapter<T>::extendRead(const std::string& fileName) const {
...
     // Read the line containing column labels and fill up the column labels
     // container.
-    auto column_labels = nextLine();
+    std::vector<std::string> column_labels{};
+    while (column_labels.size() == 0) { // keep going down rows to find labels
+        column_labels = nextLine();
+        // for labels we never expect empty elements, so remove them
+        IO::eraseEmptyElements(column_labels);
+        ++line_num;
+    }
```
[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/DelimFileAdapter.h]: If I don't do this I cannot distinguish a line with blanks from a line with labels, since I will interpret the line to be empty labels. We would have  to make our format stricter to say there cannot be blank lines between the header and the data, but I believe we have files that do have empty lines, or some random amount of tabs or spaces.

File: OpenSim/Common/TRCFileAdapter.cpp [hunk 11]
```diff
@@ -1,9 +1,11 @@
 #include "TRCFileAdapter.h"
+#include <OpenSim/Common/IO.h>
 #include <fstream>
 #include <iomanip>
 
 namespace OpenSim {
 
+const std::string _headerDelimiters{ " \t\r" };
```
[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/TRCFileAdapter.cpp]: There were some TRC files that had space as a delimiter in the header only. I don't like this, and Mokka doesn't do this, but I don't know how the TRC files in our examples were created. It is highly likely that users using our example files will have the same issue.

File: OpenSim/Common/Test/testTRCFileAdapter.cpp [hunk 12]
```diff
@@ -45,50 +45,76 @@ void compareFiles(const std::string& filenameA,
...
-                }
+            if (tokensA[i] == tokensB[i]) {
+                continue;
+            }
+            std::string tokenA{ tokensA[i] };
+            std::string tokenB{ tokensB[i] };
+            OpenSim::IO::TrimWhitespace(tokenA);
+            OpenSim::IO::TrimWhitespace(tokenB);
+            tokenA = OpenSim::IO::Lowercase(tokenA);
+            tokenB = OpenSim::IO::Lowercase(tokenB);
```
[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/Test/testTRCFileAdapter.cpp]: Yes this was mainly intended for the NaN case.  I didn't think preserving case would be an issue since the only code that can changes case is for NaNs.  I can make it more explicit for NaNs.

File: OpenSim/Common/Test/testTRCFileAdapter.cpp [hunk 13]
```diff
@@ -45,50 +45,76 @@ void compareFiles(const std::string& filenameA,
...
+            if (tokenA == tokenB) {
+                continue;
+            }
+            // We interpreted blank as NaN now revert
+            // to compare to the original file with blanks
+            if (tokenB == "nan") {
+                tokenB = "";
+            }
+            else if (tokenA == "nan") {
```
[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/Test/testTRCFileAdapter.cpp]: > Is it possible for both tokenA and tokenB to be "nan"?

No. The first `if (tokenA == tokenB)` should ensure that it is not both.

File: OpenSim/Common/Test/test.sto [hunk 14]
```diff
@@ -5,6 +5,6 @@ endheader
 		
 		  
  
-time	v1 v2
+time	v1	v2
 1.	10	20.
```
[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/Test/test.sto]: I could find any other case were we use space(s) only. Most of our .sto files have tab and multiple spaces (I have no idea why) and one cannot use space as delimiter. The previous implementation was treating all contiguous white space as a single delimiter.

File: OpenSim/Common/FileAdapter.cpp (hunk 9 above)
Comment (carmichaelong) [reply OpenSim/Common/FileAdapter.cpp]: This if statement also stuck out to me. It seems like it's not needed because of @chrisdembia's documentation, which also says
> If no matches are found, the function returns string::npos.

and `string::npos` is the greatest value of that type.

**Comment by carmichaelong** [comment]
It looks like Chris already covered a lot of what I found, so as long as those are addressed, it looks pretty good to me.

Just one other comment regarding `dataWithBlanksForMissingMarkers.trc`: Maybe it's hard to tell from looking at the file on Github alone, but it looks like the missing markers only ever appear after markers that were not missing. Perhaps there should be at least one line with missing markers interspersed between other markers?

File: OpenSim/Common/DelimFileAdapter.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (chrisdembia) [reply OpenSim/Common/DelimFileAdapter.h]: Thank you for explaining. However, if `no-empty-column-labels` is an invariant, we should make sure it's enforced properly (so that we can read files we write), and we should also document this.

Do you think it's more likely for files out in the wild to (A) have empty lines between `endheader` and the column labels or (B) have a few empty column labels?

[the remark this answers is not in the document]
Comment (aseth1) [reply OpenSim/Common/DelimFileAdapter.h]: I have seen cases of (A) and none for (B). I don't believe the previous parsing machinery was capable of processing B since contiguous delimiters  were being treated as one.

I have added enforcement to DataTable such that column labels cannot be empty and cannot contain delimiter characters or have leading and trailing spaces.

**Comment by aseth1** [comment]
> Just one other comment regarding dataWithBlanksForMissingMarkers.trc: Maybe it's hard to tell from looking at the file on Github alone, but it looks like the missing markers only ever appear after markers that were not missing. Perhaps there should be at least one line with missing markers interspersed between other markers?

That just happened to be the nature of the actual mocap data, which has the markers coming into the mocap volume.  We do have [`exampleFormat.trc`](https://github.com/opensim-org/opensim-core/pull/2187/files#diff-a1bc6aa8867f717ee2144bcb24f7deaeR1) that has a case similar to what you describe.

**Comment by aseth1** [comment]
@chrisdembia @carmichaelong the PR is ready for re-review

File: OpenSim/Common/AbstractDataTable.cpp [hunk 15]
```diff
@@ -132,28 +132,21 @@ AbstractDataTable::setColumnLabel(const size_t columnIndex,
...
-                     static_cast<unsigned>(oldLabels.size() - 1));
-
-    for(unsigned i = 0; i < oldLabels.size(); ++i) {
-        if(i == columnIndex) {
-            newLabels.upd().push_back(Value<string>(columnLabel));
-        } else {
-            auto value = Value<string>(oldLabels[i].getValue<string>());
-            newLabels.upd().push_back(value);
-        }
+    if(columnIndex == labels.size()) {
```
Comment (chrisdembia) [inline OpenSim/Common/AbstractDataTable.cpp]: I think this is a change in the behavior. Doesn't this allow having more column labels than there are columns?

File: OpenSim/Common/AbstractDataTable.h [hunk 16]
```diff
@@ -376,14 +382,30 @@ class OSIMCOMMON_API AbstractDataTable {
...
+            validateDependentsMetaData();
+        }
+        catch (const InvalidColumnLabel&) {
+            // undo any partial column label changes
+            // and restore to previous column labels if there were any
+            if (oldLabels) {
+                _dependentsMetaData.removeValueArrayForKey("labels");
+                _dependentsMetaData.setValueArrayForKey("labels", *oldLabels);
+            }
+            throw;
```
Comment (chrisdembia) [inline OpenSim/Common/AbstractDataTable.h]: +1 for undoing the change.

File: OpenSim/Common/DataTable.h [hunk 17]
```diff
@@ -1480,16 +1481,38 @@ class DataTable_ : public AbstractDataTable {
...
+        if (!_dependentsMetaData.hasKey("labels")) {
             OPENSIM_THROW(MissingMetaData, "labels");
         }
 
+        const auto labels = getColumnLabels();
+        numCols = labels.size();
+
+        // validate each label individually
+        for (const auto& label : labels) {
+            OPENSIM_THROW_IF(label.size() < 1,
```
Comment (chrisdembia) [inline OpenSim/Common/DataTable.h]: `label.empty()`?

File: OpenSim/Common/Test/testTRCFileAdapter.cpp [hunk 18]
```diff
@@ -86,22 +88,20 @@ void compareFiles(const std::string& filenameA,
...
+            tokenA = OpenSim::IO::Lowercase(tokenA);
+            tokenB = OpenSim::IO::Lowercase(tokenB);
+            // We interpreted blank as NaN now make sure
+            // to compare to interpret original file with blanks
+            if (tokenB == "") {
+                tokenB = "nan";
             }
-            if (tokenA == tokenB) {
-                continue;
+            else if (tokenA == "") {
```
Comment (chrisdembia) [inline OpenSim/Common/Test/testTRCFileAdapter.cpp]: Sorry if I  missed your response; why do you expect only `tokenA` or `tokenB` to be empty but not both?

Comment (chrisdembia) [reply OpenSim/Common/Test/testTRCFileAdapter.cpp]: Sorry; ignore this.

File: OpenSim/Common/AbstractDataTable.cpp (hunk 15 above)
Comment (aseth1) [reply OpenSim/Common/AbstractDataTable.cpp]: Funny, I thought that was it was doing previously so went out of my way to add it.
I was clearly wrong since the previous `newLabels.upd().push_back()`  was limit to i < labels.size(). I'll revert this.

**Comment by aseth1** [comment]
Thank you @chrisdembia. @carmichaelong please let me know if your concerns have been addressed.

**Comment by aseth1** [comment]
@carmichaelong?

**Comment by carmichaelong** [comment]
Sorry @aseth1 I think the Github notification got lost in the jumble of the email 😞. LGTM.

**Comment by aseth1** [comment]
Thanks @carmichaelong and @chrisdembia for the review and helpful comments.
